tools/secret-rotation/Azure.Sdk.Tools.SecretRotation.Azure/CachingTokenCredential.cs

```diff
@@ -0,0 +1,92 @@
+using System.Collections.Concurrent;
+using Azure.Core;
+using Microsoft.Extensions.Logging;
+
+namespace Azure.Sdk.Tools.SecretRotation.Azure;
+
+public class CachingTokenCredential : TokenCredential
```

Comment:
https://learn.microsoft.com/en-us/dotnet/api/overview/azure/identity-readme?view=azure-dotnet#define-a-custom-authentication-flow-with-chainedtokencredential

https://learn.microsoft.com/en-us/javascript/api/@azure/identity/chainedtokencredential?view=azure-node-latest

Reply:
lol... I thought for sure I copied the link to those docs. I must have gotten distracted and forget.

Reply:
ChainedTokenCredential doesn't contain any caching logic.  It just forwards the token request to each TokenCredential in `_sources` until one returns a token without throwing a CredentialUnavailableException

https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/src/Credentials/ChainedTokenCredential.cs

```cs
foreach (TokenCredential source in _sources)
{
    try
    {
        AccessToken token = async
            ? await source.GetTokenAsync(requestContext, cancellationToken).ConfigureAwait(false)
            : source.GetToken(requestContext, cancellationToken);
        groupScopeHandler.Dispose(default, default);
        return token;
    }
    catch (CredentialUnavailableException e)
    {
        exceptions.Add(e);
    }
    catch (Exception e) when (!cancellationToken.IsCancellationRequested)
    {
        throw new AuthenticationFailedException(AuthenticationFailedErrorMessage + e.Message, e);
    }
}
```

Reply:
@schaabs is there no existing way to cache a credential?

Reply:
Our dev time credentials such as the `AzureCliCredential` don't do any caching of tokens as of now. It is a feature we can add now as the MSAL team has given us a generic mechanism to cache tokens from any source (which we used to add caching to the `ManagedIdentityCredential`. 

If you're only using the credential with clients from the Azure SDK, then caching at the credential level is not really necessary as the clients themselves cache the tokens they acquire. However, if you're calling `GetToken(Async)` directly or using a client outside of the Azure SDK, it's likely that each call will spawn a CLI process.

If you do need caching for this, I would prefer you file a feature request on the Identity library where we can use the same MSAL caching we use for other credentials as the logic is more complex than it might seem.

Reply:
@hallipr given that I suggest we live with the performance issue for now and file the feature request instead of trying to maintain this cache ourselves.

Reply:
sounds good to me